### PR TITLE
Feature/user boundary filter

### PR DESCRIPTION
**Pull request by flibbertigibbet**
Filter map view by user drawn boundary.

Record list does not yet filter by user boundary, and the map view still does not filter by anything else, other than record type. Those things can go in separate PRs.

Companion PR on ashlar for the boundary filter back-end and endpoint used to get SQL to pass along to Windshaft: azavea/ashlar/pull/47

Also bumps memory for app Vagrant box.

### CODE REVIEW

**Comment by flibbertigibbet** [comment]
image

File: windshaft/driver.js [hunk 1]
```diff
@@ -86,18 +86,20 @@ function getRequestParameters(request) {
         params.interactivity = 'uuid,occurred_from,data';
         params.style = eventsStyle;
 
-        // build query for record points
-        params.sql = baseRecordQuery;
-        if (params.id !== 'ALL') {
-            params.sql += filterRecordQuery + params.id + "'";
+        // build query for record points if do not already have a query
+        if (request.query.sql) {
+            params.sql = request.query.sql;
```
Comment (sharph) [inline windshaft/driver.js]: Won't this allow an attacker (who has authenticated) to execute arbitrary SQL against the database?

Comment (sharph) [inline windshaft/driver.js]: Maybe sign the SQL with a combination of [itsdangerous](http://pythonhosted.org/itsdangerous/) and [node-jsjws](https://github.com/davedoesdev/node-jsjws)?

Comment (flibbertigibbet) [inline windshaft/driver.js]: Windshaft wraps all queries in a `SELECT`, which presumably would prevent actions other than reading. See #144, which should also add some controls via database roles. Windshaft is designed to allow arbitrary SQL queries.

**Comment by sharph** [comment]
You can escape the `SELECT` wrapper with semicolons as demonstrated in https://github.com/WorldBank-Transport/open-transit-indicators/issues/360

I think #144 would be a prerequisite for accepting this PR if we're to go that route. Would there be any concerns about different orgs accessing each others data or is that out of scope at this point?

**Comment by flibbertigibbet** [comment]
And WorldBank-Transport/open-transit-indicators#360 was closed without a fix. SQL vuln, inherent to Windshaft, is a larger issue that we'll have to address separately.

**Comment by sharph** [comment]
https://github.com/WorldBank-Transport/open-transit-indicators/issues/360 was closed in https://github.com/WorldBank-Transport/open-transit-indicators/pull/413 though it wasn't properly linked.

I strongly disagree that there is a SQL vulnerability or problem in windshaft. winshaft assumes it is being given trusted data and we are breaking that assumption by passing it untrusted user submitted data.

**Comment by flibbertigibbet** [comment]
CartoDB takes arbitrary, user-submitted SQL without any filtering or testing, which is passed straight to the SQL API and Windshaft. The only checks are via user/role permissions.

We need to implement custom, complex filters here, involving unknown fields, so limiting the options by field as we've done in the past won't work in this case. The SQL here is *not* from the user, it is from Django, via the app frontend. We don't have anything in place yet to prevent requests that aren't from the app going to Windshaft.

**Comment by sharph** [comment]
Whether the request came *from the user* is an issue of semantics. The request was sent from the user's machine and should be considered untrusted. CartoDB implements access control at the database level with user permissions as you say, and *at the very least* we should lock down windshaft's user to only `SELECT`. Preferably limiting by field could be implemented, perhaps by signing the query as I've suggested earlier.

I'm afraid that implementing the feature without the relevant access controls in this case will open a security hole that will go forgotten.

**Comment by flibbertigibbet** [comment]
There's an open issue for it I've created, so it will not be forgotten. We have not done the cards yet for user permissions, so what should have access is an open question. There is no sensitive data on staging, nor any significant negative consequences for data loss or tampering there. This PR is however needed for further work to be done, and security is again a larger issue best addressed separately, for completeness.

**Comment by sharph** [comment]
I'll +1 with lots of hesitation and alarm flags raised to keep things moving along. Please consider adding a user and role with `SELECT` access to the relevant tables for windshaft -- it is not blocked by the user permission card and will at least keep someone from dropping tables or reading/`UPDATE`ing the users table.

**Pull request merged**